File: fast_mm_matlab/src/utils/PrepareShapes.m
% Given a level, list all shapes (i, j, k) where i + j + k = 2^{level}.
% Return a n*3 matrix.

function shapes = PrepareShapes(level)
  shapes = [];
  sum_col = 2 ^ level;
  for i = 0 : sum_col
    for j = 0 : sum_col - i
      k = sum_col - i - j;
      shapes = [shapes; i, j, k];
    end
  end
end

File: fast_mm_python/src/utils/PrepareShapes.py
"""
Given a level, list all shapes (i, j, k) where i + j + k = 2^level.
"""

import numpy as np


def PrepareShapes(level):
    """
    Given a level, list all shapes (i, j, k) where i + j + k = 2^level.
    
    Args:
        level: Level value
    
    Returns:
        shapes: Array of shape (n, 3) containing all valid triples
    """
    shapes = []
    sum_col = 2 ** level
    
    for i in range(sum_col + 1):
        for j in range(sum_col - i + 1):
            k = sum_col - i - j
            shapes.append([i, j, k])
    
    return np.array(shapes, dtype=int)

